zml/tokenizer/homemade.zig

```diff
@@ -0,0 +1,1215 @@
+    }
+
+    /// Adds a new token (without copying it)
+    pub fn addOwnedToken(self: *Tokenizer, score: f32, token: []const u8) void {
+        const i = self.next_token_id;
+        std.debug.assert(i < self.vocab_size);
+        self.next_token_id += 1;
+
+        self.scores[i] = score;
+        self.tokens[i] = token;
+        const v = self.token_lookup.getOrPutAssumeCapacity(token);
+        if (!v.found_existing) {
+            v.value_ptr.* = i;
+        }
+    }
+
+    pub fn addOwnedTokenByIndex(self: *Tokenizer, i: u32, score: f32, token: []const u8) void {
+        std.debug.assert(i < self.vocab_size);
+        self.next_token_id += 1;
+        self.scores[i] = score;
+        self.tokens[i] = token;
+        const v = self.token_lookup.getOrPutAssumeCapacity(token);
+        if (!v.found_existing) {
+            v.value_ptr.* = @intCast(i);
+        }
+    }
+
+    pub fn lookup(self: *const Tokenizer, str: []const u8) ?u32 {
+        return self.token_lookup.get(str);
+    }
+
+    pub fn tokenToId(self: *const Tokenizer, token: []const u8) ?u32 {
+        return self.token_lookup.get(token);
+    }
+
+    pub const EncodeOptions = struct {
+        /// Should the beginning of sentence '<s>' token be added.
+        add_bos: bool = true,
+        add_eos: bool = false,
+        pad_to: u32 = 0,
+        // Print tokenization intermediary steps.
+        debug: bool = false,
+    };
+
+    pub fn encode(self: *const Tokenizer, allocator: std.mem.Allocator, raw: []const u8, options: EncodeOptions) ![]u32 {
+        if (options.debug) log.debug("Tokenizer.encode('{s}')", .{raw});
+        const input = if (self.normalizer) |n| try n.normalize(allocator, raw) else raw;
+        defer if (self.normalizer) |_| allocator.free(input);
+        if (options.debug) log.debug("Tokenizer.encode.normalize -> '{s}'", .{input});
+
+        // Allocate a buffer that can fit all indices as well as extra character if requested.
+        // We then slice it so that the token merging code doesn't see the bos token.
+        const tok_buff_alloc = try allocator.alloc(u32, @max(options.pad_to, input.len + 2));
+        const tok_buff = if (options.add_bos) tok_buff_alloc[1..] else tok_buff_alloc;
+
+        const MergeState = union(enum) { ready: u32, nope, hard_space, idk };
+        const mergeable = try allocator.alloc(MergeState, tok_buff.len);
+
+        var num_tokens: usize = 0;
+        var it: CharTokenIterator = .{ .input = input };
+        while (try it.nextCodepointToken(self)) |token| : (num_tokens += 1) {
+            tok_buff[num_tokens] = token;
+            mergeable[num_tokens] = if (token == self.special_tokens.hard_space)
+                .hard_space
+            else
+                .idk;
+        }
+
+        var stable_prefix: usize = 0;
+        var stable_off: usize = 0;
+        while (true) {
+            // This code is a bit overcomplicated cause I'm abstracting over two algorithms:
```

Comment:
this is a restore of previous tokenizer.zig we have plus:
* a disclaimer at the top 
* this note in the middle
* binding to the encoder/decoder api